File: csharp/Kiahk/Feast.cs
using System.Collections.Generic;

namespace Kiahk;

/// <summary>
/// A calendar-resolved feast: a <see cref="FeastRecord"/>'s metadata paired with the
/// Gregorian date on which it falls for a particular year.
/// </summary>
public sealed record Feast(
    string Id,
    string Type,
    string Category,
    IReadOnlyDictionary<string, string> Names,
    GregorianDate GregorianDate)
{
    /// <summary>
    /// Return the feast's localized name for <paramref name="locale"/>.
    /// Supported locales: "en", "ar". Unknown locales throw <see cref="UnsupportedLocaleException"/>.
    /// </summary>
    public string Name(string locale)
    {
        if (Names.TryGetValue(locale, out var n))
        {
            return n;
        }
        throw new UnsupportedLocaleException(Id, locale);
    }
}

File: swift/Sources/Kiahk/Feast.swift
import Foundation

/// A calendar-resolved feast: a `FeastRecord`'s metadata paired with the
/// Gregorian date on which it falls for a particular year.
public struct Feast: Equatable, Hashable, Sendable {
    public let id: String
    public let type: String            // "fixed" | "moveable"
    public let category: String        // "major" | "minor"
    public let names: [String: String]
    public let gregorianDate: GregorianDate

    public init(
        id: String,
        type: String,
        category: String,
        names: [String: String],
        gregorianDate: GregorianDate
    ) {
        self.id = id
        self.type = type
        self.category = category
        self.names = names
        self.gregorianDate = gregorianDate
    }

    /// Return the feast's localized name for `locale`. Supported: "en", "ar".
    /// Unknown locales throw `KiahkError.unsupportedLocale`.
    public func name(locale: String) throws -> String {
        guard let n = names[locale] else {
            throw KiahkError.unsupportedLocale(feastID: id, locale: locale)
        }
        return n
    }
}
